src/staking/DyadXPv2.sol

```diff
@@ -0,0 +1,203 @@
+    /// to spend on behalf of `owner`
+    function allowance(address, address) external pure returns (uint256) {
+        return 0;
+    }
+
+    /// @notice Sets `amount` as the allowance of `spender` over the caller's tokens.
+    /// @dev Be aware of front-running risks: https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
+    function approve(address, uint256) external pure returns (bool) {
+        revert ApproveNotAllowed();
+    }
+
+    /// @notice Moves `amount` tokens from `from` to `to` using the allowance mechanism.
+    /// `amount` is then deducted from the caller's allowance.
+    function transferFrom(address, address, uint256) external pure returns (bool) {
+        revert TransferNotAllowed();
+    }
+
+    function beforeKeroseneWithdrawn(uint256 noteId, uint256 amountWithdrawn) external {
+        if (msg.sender != address(VAULT_MANAGER)) {
+            revert NotVaultManager();
+        }
+
+        NoteXPData memory lastUpdate = noteData[noteId];
+
+        uint256 xp = _computeXP(lastUpdate);
+
+        uint256 slashedXP = xp.mulDivUp(amountWithdrawn, lastUpdate.keroseneDeposited);
+
+        if (slashedXP > xp) {
+            slashedXP = xp;
+        }
+
+        uint120 newXP = uint120(xp - slashedXP);
+
+        noteData[noteId] = NoteXPData({
+            lastAction: uint40(block.timestamp),
+            keroseneDeposited: uint96(lastUpdate.keroseneDeposited - amountWithdrawn),
+            lastXP: newXP,
+            totalXP: xp,
+            dyadMinted: DYAD.mintedDyad(noteId)
+        });
+
+        if (newXP > lastUpdate.lastXP) {
+          emit Transfer(address(0), DNFT.ownerOf(noteId), newXP - lastUpdate.lastXP);
+        } else {
+          emit Transfer(DNFT.ownerOf(noteId), address(0), lastUpdate.lastXP - newXP);
+        }
+    }
+
+    function updateXP(uint256 noteId) external {
+        if (msg.sender != address(VAULT_MANAGER)) {
+            revert NotVaultManager();
+        }
+
+        NoteXPData memory lastUpdate = noteData[noteId];
+        uint256 newXP = _computeXP(lastUpdate);
+
+        noteData[noteId] = NoteXPData({
+            lastAction: uint40(block.timestamp),
+            keroseneDeposited: uint96(KEROSENE_VAULT.id2asset(noteId)),
+            lastXP: uint120(newXP),
+            totalXP: lastUpdate.totalXP,
+            dyadMinted: DYAD.mintedDyad(noteId)
+        });
+
+        emit Transfer(address(0), DNFT.ownerOf(noteId), newXP - lastUpdate.lastXP);
+    }
+
+    function _authorizeUpgrade(address) internal view override onlyOwner {}
+
+    function _computeXP(NoteXPData memory lastUpdate) internal view returns (uint256) {
+        uint256 elapsed = block.timestamp - lastUpdate.lastAction;
+        uint256 deposited = lastUpdate.keroseneDeposited;
+        uint256 dyadMinted = lastUpdate.dyadMinted;
+        uint256 totalXP = lastUpdate.totalXP;
+
+        uint256 accrualRateModifier = totalXP > 0 ? 1e18 / totalXP.log10() : 1e18;
```

Comment:
does this work given any elapsed time (t)

for instance given two notes, each with the same base accrual rate, does claiming at different frequencies net them different total XP accrued?

Reply:
not sure I understand what you mean

Reply:
its a good point. seems like we need to modify our equations.